core/src/main/java/com/scalar/db/transaction/consensuscommit/PrepareMutationComposer.java

```diff
@@ -137,28 +160,35 @@ private void add(Get base) {
     mutations.add(put);
   }
 
-  private List<Value<?>> createBeforeValues(Mutation base, TransactionResult result) {
-    List<Value<?>> values = new ArrayList<>();
+  private List<Column<?>> createBeforeColumns(Mutation base, TransactionResult result) {
+    List<Column<?>> columns = new ArrayList<>();
     result
-        .getValues()
+        .getColumns()
         .values()
         .forEach(
-            v -> {
-              if (isBeforeRequired(v, base.getPartitionKey(), base.getClusteringKey())) {
-                values.add(v.copyWith(Attribute.BEFORE_PREFIX + v.getName()));
+            column -> {
+              if (isBeforeRequired(column, base.getPartitionKey(), base.getClusteringKey())) {
+                if (column.getName().equals(Attribute.VERSION) && column.hasNullValue()) {
+                  // For preparing an update of a deemed-commit state record, we need to use
+                  // version 0 rather than NULL since we want to distinguish the preparation from
+                  // an initial record insertion.
+                  columns.add(IntColumn.of(Attribute.BEFORE_VERSION, 0));
```

Comment:
As commented in the code, we don't use the original NULL value for `before_version` in prepare records
to distinguish them from prepare records for newly-inserted ones.

Reply:
Thanks for seeking clarification. Simply put, the record cannot be rollbacked correctly since the recovery process falls in the deletion path for the initial record case here.
https://github.com/scalar-labs/scalardb/blob/b3cf9c9735c534a37404a6366734710d0c79c3fc/core/src/main/java/com/scalar/db/transaction/consensuscommit/RollbackMutationComposer.java#L67-L69

Reply:
Ah, right. Thank you!
I think it would be great if you could make the comment more specific about the case.
(i.e., adding the reason why we want to use 0 instead of null for before_version.)

Reply:
Sure, I will try to revise it with a more specific reason.

Reply:
@feeblefakie Fixed in d4566f0. PTAL when you get a chance.